test/Main.hs

```diff
@@ -100,18 +95,23 @@ parseResponses :: TestTree
       , (JSON.object [("foo", JSON.object empty)], "Invalid status code pattern: \"foo\"")
       ]
 
-    mkAccepted :: (JSON.Value, Paths.Responses) -> TestTree
-    mkAccepted (input, expected)
-      = testCase (convertText ("accepts: " <> show input))
-      $ pure expected @=? parseJSON input
-
-    mkRejected :: (JSON.Value, Text) -> TestTree
-    mkRejected (input, expectedMessage)
-      = testCase (convertText ("rejects: " <> show input))
-      $ Left ("Error in $: " <> convertText expectedMessage) @=? parseJSON input
-
-    parseJSON :: JSON.Value -> Either String Paths.Responses
-    parseJSON = JSON.parseEither JSON.parseJSON
-
 main :: IO ()
 main = defaultMain suite
+
+mkAccepted
+  :: (Eq a, JSON.FromJSON a, Show a)
+  => (JSON.Value, a)
+  -> TestTree
+mkAccepted (input, expected)
+  = testCase (convertText ("accepts: " <> show input))
+  $ pure expected @=? JSON.parseEither JSON.parseJSON input
+
+mkRejected
+  :: (Eq a, Show a)
+  => (JSON.Value -> JSON.Parser a)
+  -> Text
+  -> (JSON.Value, Text)
+  -> TestTree
+mkRejected parseJSON prefix (input, message)
+  = testCase (convertText ("rejects: " <> show input))
```

Comment:
Instance of https://github.com/mbj/openapi/pull/9/commits/9a21e870c76224a3862b25c61d99940fb8a89ce0#r324255370